include/irr/asset/IVirtualTexture.h

```diff
@@ -0,0 +1,750 @@
+    {
+        if (_initSharedResources)
+        {
+            uint32_t pgtabSzSqr = (1u<<_pgTabSzxy_log2);
+            pgtabSzSqr *= pgtabSzSqr;
+            const size_t spacePerAllocator = pg_tab_addr_alctr_t::reserved_size(pgtabSzSqr, pgtabSzSqr, 1u);
+            m_pgTabAddrAlctr_reservedSpc = reinterpret_cast<uint8_t*>( _IRR_ALIGNED_MALLOC(spacePerAllocator*_pgTabLayers, _IRR_SIMD_ALIGNMENT) );
+            for (uint32_t i = 0u; i < _pgTabLayers; ++i)
+            {
+                auto& alctr = (*m_pageTableLayerAllocators)[i];
+                alctr = pg_tab_addr_alctr_t(m_pgTabAddrAlctr_reservedSpc+i*spacePerAllocator, 0u, 0u, pgtabSzSqr, pgtabSzSqr, 1u);
+            }
+        }
+
+        if (_initSharedResources)
+        {
+            decltype(m_freePageTableLayerIDs)::container_type stackBackend;
+            stackBackend.resize(_pgTabLayers);
+            for (uint32_t i = 0u; i < _pgTabLayers; ++i)
+                stackBackend[i] = _pgTabLayers-1u-i;
+            m_freePageTableLayerIDs = decltype(m_freePageTableLayerIDs)(std::move(stackBackend));
+        }
+        {
+            const uint32_t pgSzLog2 = getPageExtent_log2();
+            bool ok = SMiptailPacker::computeMiptailOffsets(m_miptailOffsets, pgSzLog2, m_tilePadding);
+            assert(ok);
+        }
+    }
+
+    virtual STextureData pack(const image_t* _img, const IImage::SSubresourceRange& _subres, ISampler::E_TEXTURE_CLAMP _wrapu, ISampler::E_TEXTURE_CLAMP _wrapv, ISampler::E_TEXTURE_BORDER_COLOR _borderColor) = 0;
+
+    virtual core::smart_refctd_ptr<image_view_t> createPageTableView() const = 0;
+
+    image_t* const & getPageTable() const { return m_pageTable.get(); }
+    uint32_t getPageTableExtent_log2() const { return m_pgSzxy_log2; }
+    uint32_t getPageExtent() const { return m_pgSzxy; }
+    uint32_t getPageExtent_log2() const { return core::findLSB(m_pgSzxy); }
+    uint32_t getTilePadding() const { return m_tilePadding; }
+    const core::stack<uint32_t>& getFreePageTableLayersStack() const { return m_freePageTableLayerIDs; }
+    core::SRange<const uint32_t> getLayerToViewIndexMapping() const { return {m_layerToViewIndexMapping->begin(),m_layerToViewIndexMapping->end()}; }
+    const auto& getResidentStorages() const { return m_storage; }
+    typename SamplerArray::range_t getFloatViews() const  { return m_fsamplers.getViews(); }
+    typename SamplerArray::range_t getIntViews() const { return m_isamplers.getViews(); }
+    typename SamplerArray::range_t getUintViews() const { return m_usamplers.getViews(); }
+
+    size_t getLayerToViewIndexLUTBytesize() const
+    {
+        return m_layerToViewIndexMapping->size()*sizeof(uint32_t);
+    }
+    void writeLayerToViewIndexLUTContents(void* _dst) const
+    {
+        memcpy(_dst, m_layerToViewIndexMapping->data(), getLayerToViewIndexLUTBytesize());
+        m_layer2viewWasUpdatedSinceLastQuery = false;
```

Comment:
maybe it would be just easier to return a constant reference to `m_layerToViewIndexMapping`

resetting this should be optional (may have to get/write multiple times)

Reply:
since you already have `getLayerToViewIndexMapping` this makes one of the functions redunant, instead I propose a single (with `m_layerToViewIndexMapping` being an std::array or a struct)
```cpp
inline const auto& getLayerToViewIndexMapping() const
{
    return m_layerToViewIndexMapping;
}
struct reset_update_t {};
const auto& getLayerToViewIndexMapping(reset_update_t reset_update)
{
    m_layer2viewWasUpdatedSinceLastQuery = false;
    return m_layerToViewIndexMapping;
}
```